lib/core/src/Cardano/Wallet/Primitive/Scripts.hs

```diff
@@ -0,0 +1,92 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- |
+-- Copyright: © 2018-2020 IOHK
+-- License: Apache-2.0
+--
+-- An implementation of wallet's verification key discovery
+-- in a script.
+
+module Cardano.Wallet.Primitive.Scripts
+    ( isShared
+    , retrieveAllVerKeyHashes
+    , toKeyHash
+    ) where
+
+import Prelude
+
+import Cardano.Address.Derivation
+    ( xpubPublicKey )
+import Cardano.Address.Script
+    ( KeyHash (..), Script (..), ScriptHash (..), toScriptHash )
+import Cardano.Crypto.Wallet
+    ( XPub )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (..)
+    , DerivationType (..)
+    , Index (..)
+    , Role (..)
+    , SoftDerivation (..)
+    , hex
+    )
+import Cardano.Wallet.Primitive.AddressDerivation.Shelley
+    ( ShelleyKey (..) )
+import Cardano.Wallet.Primitive.AddressDiscovery.Sequential
+    ( SeqState (..), accountPubKey )
+import Crypto.Hash.Utils
+    ( blake2b224 )
+
+import qualified Data.Map.Strict as Map
+import qualified Data.Text.Encoding as T
+
+instance Ord ScriptHash where
+    compare (ScriptHash sh1) (ScriptHash sh2) =
+        compare (T.decodeUtf8 $ hex sh1) (T.decodeUtf8 $ hex sh2)
```

Comment:
What's wrong with the stock `Ord` instance? Why comparing base16 representations :thinking:  ?

Reply:
there is no `Ord` instance for `ScriptHash` and `HashKey` in cardano-addresses, unfortunately.

Reply:
Okay, but one can still use `StandaloneDeriving` and write: `derive instance Ord ScriptHash`. Although we should really get those instances in cardano-addresses.

Reply:
indeed, done in https://github.com/input-output-hk/cardano-wallet/pull/2308/commits/91f8cc5fd87e6d57ebfca5d51d84c3eb8418a439